hack/bin/Sbom.hs

```diff
@@ -0,0 +1,376 @@
+  MkSBomMeta
+    { drvPath = drvPath,
+      outPath = Identity outPath,
+      directDeps = Identity directDeps,
+      description = Nothing,
+      homepage = Nothing,
+      licenseSpdxId = [],
+      name = Nothing,
+      typ = Nothing,
+      urls = [],
+      version = Nothing
+    }
+
+data SBomMeta f = MkSBomMeta
+  { drvPath :: Text,
+    description :: Maybe Text,
+    homepage :: Maybe Text,
+    licenseSpdxId :: [Maybe License],
+    name :: Maybe Text,
+    typ :: Maybe Text,
+    urls :: [Maybe Text],
+    version :: Maybe Text,
+    outPath :: f Text,
+    directDeps :: f [Text]
+  }
+
+deriving stock instance (Eq (f [Text]), Eq (f Text)) => Eq (SBomMeta f)
+
+deriving stock instance (Ord (f [Text]), Ord (f Text)) => Ord (SBomMeta f)
+
+deriving stock instance (Show (f [Text]), Show (f Text)) => Show (SBomMeta f)
+
+type Meta = SBomMeta Proxy
+
+instance FromJSON Meta where
+  parseJSON (Object val) =
+    MkSBomMeta
+      <$> do val .: "drvPath"
+      <*> do val .: "description"
+      <*> do val .: "homepage"
+      <*> do val .: "licenseSpdxId"
+      <*> do val .: "name"
+      <*> do val .: "type"
+      <*> do val .: "urls"
+      <*> do val .: "version"
+      <*> pure Proxy
+      <*> pure Proxy
+  parseJSON invalid = typeMismatch "Object" invalid
+
+type SBom = Map Text (SBomMeta Identity)
+
+type MetaDB = Map Text (SBomMeta Proxy)
+
+type ClosureInfo = Tree ByteString
+
+type PathInfo = [(Text, (Text, [Text]))]
+
+data Visit a = Seen a | Unseen a
+  deriving stock (Eq, Ord, Show)
+
+data SerializeSBom = MkSerializeSBom
+  { -- | the version of the SBom; this is version of the old SBom + 1
+    sbom'version :: Natural,
+    -- | name of the component the SBom is generated for
+    sbom'component :: Text,
+    -- | the creator of the component the SBom is generated for
+    sbom'manufacture :: Text,
+    -- | the supplier (manufacturer or repackager or distributor)
+    sbom'supplier :: Maybe Text,
+    -- | (spdxids of) licenses of the product
+    sbom'licenses :: [Text]
+  }
+
+defaultSerializeSBom :: SerializeSBom
+defaultSerializeSBom =
+  MkSerializeSBom
+    { sbom'version = 1,
+      sbom'component = "wire-server",
+      sbom'manufacture = "wire",
+      sbom'supplier = Nothing,
+      sbom'licenses = ["AGPL-3.0-or-later"]
+    }
+
+-- FUTUREWORK(mangoiv): we can also have
+--
+-- - qualifiers: extra qualifying data for a package such as an OS, architecture, a distro, etc. Optional and type-specific.
+-- - subpath: extra subpath within a package, relative to the package root. Optional.
+-- - use heuristics based approach to finding original repositories for packages, e.g. pkg:hackage....
+mkPurl :: SBomMeta Identity -> Text
+mkPurl meta =
+  mconcat
+    [ "pkg:",
+      repo,
+      "/",
+      fromMaybe (runIdentity meta.outPath) meta.name,
+      maybe "" ("@" <>) meta.version
+    ]
+  where
+    repo
+      | any (maybe False (T.isInfixOf "hackage.haskell.org")) meta.urls = "hackage"
```

Comment:
Is this really the expected package format: `pkg:hackage/svens-package@.0.0.42`?
I mean: You won't find anything with this URL on the internet, right? :thinking: